decimal.go

```diff
@@ -2032,6 +2026,31 @@ func RescalePair(d1 Decimal, d2 Decimal) (Decimal, Decimal) {
 	return d1, d2
 }
 
+// safeFactorial calculates the factorial of a given number and caches the result.
+func safeFactorial(i int64) Decimal {
+	factorialsMu.RLock()
+	if len(factorials) >= int(i) && !factorials[i-1].IsZero() {
+		factorialsMu.RUnlock()
+		return factorials[i-1]
+	}
+	factorialsMu.RUnlock()
+
+	factorialsMu.Lock()
+	defer factorialsMu.Unlock()
```

Comment:
Shouldn't we lock at line 2046? I think `factorialsMu.RLock()` in line 2031 is sufficient for the line 2043 check as there are no writes

Reply:
No, we need a write lock at this point, 2043 and 2032 if's are identical. The second if is just a safety measure between the times we unlock the mutex.

Reply:
> No, we need a write lock at this point

But you don't write until line 2049, no? 2043 and 2044 don't need write mutex?

Reply:
how are you going to upgrade from read to write mutex without a race condition?

Reply:
Place `RUnlock` before return in 2044 and 2046, and then `Lock` before 2047

Reply:
And you get yourself another race condition.